### PR TITLE
Support indexer with multiple arguments

### PR DESCRIPTION
**Pull request by idigra**
This PR contains support in multiple arguments.

1. Fix null reference exception in case of multiple arguments indexer.
2. Proposal for new API method SetupSetMethod which allows using It.IsAny for the indexer arguments (can be used also for properties, although existing API already covers properties). This is only in proposal level, should be enhanced (documentation, overloading for any count of indexer arguments, exceptions on illegal input).

### CODE REVIEW

**Comment by stakx** [comment]
Hi @idigra, and thanks for this one, too. Same as with the other one, I'll try to find some time for reviewing this sometime during the next few days.

I can say right now, however, that it shouldn't be necessary to introduce a `SetupSetMethod`. You'll find that the existing `Setup`, `SetupGet`, `SetupSet` etc. methods already try to recognize when they are called on indexers&mdash;indexers shouldn't have their own dedicated methods, from a user's perspective that doesn't make much sense. (`SetupGet` and `SetupSet` are probably already confusing enough.)

**Comment by idigra** [comment]
I now see that the second issue I opened (https://github.com/moq/moq4/issues/696, mentioned above) is a duplicate of https://github.com/moq/moq4/issues/218 which is closed in favor of https://github.com/moq/moq4/issues/414. I couldn't find the reason why the SetupSet overload you mentioned was deprecated. Also consider the solution I proposed in my issue https://github.com/moq/moq4/issues/696.

**Comment by stakx** [comment]
> I couldn't find the reason why the SetupSet overload you mentioned was deprecated.

I don't know the exact reason, either. I am guessing that this decision was made simply because being able to write `SetupSet(m => m.Something = value)` is nicer than `SetupSet(m => m.Something, () => value)`.

File: src/Moq/Mock.cs [hunk 1]
```diff
@@ -664,13 +664,12 @@ private static VoidSetupPhrase<T> SetupSetPexProtected<T>(Mock<T> mock, Action<T
 
 					matchers[valueIndex] = new MatchExpression(last.Match);
 
-					if (arguments.Length == 2)
+					for (int i = 0; i < arguments.Length - 1; i++)
```
Comment (stakx) [inline src/Moq/Mock.cs]: I think it would be clearer to compare `i < valueIndex` to emphasize the fact that we're excluding the `value` parameter.

File: tests/Moq.Tests/CallbacksFixture.cs [hunk 2]
```diff
@@ -474,6 +474,105 @@ public void CallbackWithRefParameterCannotModifyNonRefParameter()
 			Assert.Equal("input", value);
 		}
 
+		[Fact]
+		public void CallbackWithIndexerGetter()
```
Comment (stakx) [inline tests/Moq.Tests/CallbacksFixture.cs]: I appreciate this test, and it would be good to have one like it (if we don't have one like it yet, I didn't double-check TBH)... but could you explain how it is related to this PR?

If it isn't directly related to this PR and you simply added it for completeness' sake, then I'd appreciate if you submitted it in a separate PR.

File: tests/Moq.Tests/CallbacksFixture.cs [hunk 3]
```diff
@@ -474,6 +474,105 @@ public void CallbackWithRefParameterCannotModifyNonRefParameter()
...
+		}
+
+		[Fact]
+		public void CallbackWithMultipleArgumentIndexerGetter()
+		{
+			int x = default(int);
+			int y = default(int);
+
+			var mock = new Mock<IFoo>();
+			mock.Setup(f => f[It.IsAny<int>(), It.IsAny<int>()])
```
Comment (stakx) [inline tests/Moq.Tests/CallbacksFixture.cs]: Same as in an earlier review comment, I appreciate the test, but it doesn't seem to be related to this PR, since it uses the expression tree-based `Setup` instead of the delegate-based `SetupSet`...?

Could you state how it relates to the bugfix in your PR (and if it doesn't, could you remove it and optionally re-submit in a separate PR?)

File: tests/Moq.Tests/CallbacksFixture.cs [hunk 4]
```diff
@@ -474,6 +474,105 @@ public void CallbackWithRefParameterCannotModifyNonRefParameter()
...
+				}));
+
+			mock.Object[3, 13] = 2;
+			Assert.Equal(3, x);
+			Assert.Equal(13, y);
+			Assert.Equal(2, result);
+		}
+
+		[Fact]
+		public void CallbackWithMultipleArgumentIndexerSetterWithAny()
```
Comment (stakx) [inline tests/Moq.Tests/CallbacksFixture.cs]: Please remove along with the rest of `SetupSetMethod`-related functionality.

File: tests/Moq.Tests/CallbacksFixture.cs [hunk 5]
```diff
@@ -490,7 +589,7 @@ public class Derived : Base
 		private void TraceMe(Base b)
 		{
 		}
-		
+
```
Comment (stakx) [inline tests/Moq.Tests/CallbacksFixture.cs]: It's a tiny detail, I know, but could you please revert this whitespace-only change?

File: src/Moq/Mock.Generic.cs [hunk 6]
```diff
@@ -344,6 +344,12 @@ public ISetup<T> SetupSet(Action<T> setterExpression)
 			return Mock.SetupSet<T>(this, setterExpression, null);
 		}
 
+		/// <include file='Mock.Generic.xdoc' path='docs/doc[@for="Mock{T}.SetupIndexerSet"]/*'/>
+		public ISetup<T> SetupSetMethod<T1, T2, TProperty>(Expression<Action<Action<T1, T2, TProperty>>> setter)
```
Comment (stakx) [inline src/Moq/Mock.Generic.cs]: Please remove.

File: src/Moq/Mock.cs [hunk 7]
```diff
@@ -683,6 +682,28 @@ private static VoidSetupPhrase<T> SetupSetPexProtected<T>(Mock<T> mock, Action<T
 			}
 		}
 
+		internal static ISetup<T> SetupSetMethod<T, TProperty>(
```
Comment (stakx) [inline src/Moq/Mock.cs]: Please remove.

File: tests/Moq.Tests/CallbacksFixture.cs (hunk 5 above)
Comment (idigra) [reply tests/Moq.Tests/CallbacksFixture.cs]: Automatically done by VS. I reverted it manually using notepad++, hope that now it's OK.

File: src/Moq/Mock.cs (hunk 1 above)
Comment (idigra) [reply src/Moq/Mock.cs]: I tried to be consistent with the for loop that is some lines up (line 635).
Let me know if you prefer to (1) leave it as I've done, for maximal consistency between the two code lines AND minimum changes, (2) Fix both for loops, for maximal consistency AND answer your request, or (3) Fix only mine, so at least new code will answer your request (in cost of consistency).

Comment (stakx) [reply src/Moq/Mock.cs]: Fair enough. Never mind for the moment, let's leave it as you put it.

**Comment by stakx** [comment]
@idigra, Moq 4.10.1, which includes your bugfix, has just been published on NuGet.

**Comment by idigra** [comment]
Thanks. I finally came to verify it, but then I understand that without extending the API to support Any matcher in indexer argument I can't really reach a scenario in which the bug was reproduced.
I would be happy to check it if you have a real scenario in which the bug occurs.

**Comment by idigra** [comment]
https://github.com/moq/moq4/issues/696
